File: backend/service/article.go
package service

import (
	"log"
	"mime/multipart"
	"net/http"
	"strconv"

	"github.com/grkon03/newblog/backend/service/repository"
	"github.com/grkon03/newblog/backend/util"
	"github.com/labstack/echo/v4"
)

type ArticleAPI struct {
	h  repository.ArticleHandler
	ih repository.ImageHandler
}

func (a *ArticleAPI) GetArticle(c echo.Context) error {
	id64, err := strconv.ParseUint(c.Param("id"), 10, 64)
	id := uint(id64)

	if err != nil {
		log.Println(err.Error())
		c.String(http.StatusBadRequest, "Bad request")
		return err
	}

	article, err := a.h.GetArticle(id)
	if err != nil {
		log.Println(err.Error())
		c.String(http.StatusNotFound, "No such article")
		return err

	}

	c.JSON(http.StatusOK, *article)
	return nil

}

// ?from=xxx&count=xxx
func (a *ArticleAPI) GetArticles(c echo.Context) error {
	var from, count uint = 0, 0

	param := c.QueryParam("from")
	if param != "" {
		from64, err := strconv.ParseInt(param, 10, 64)
		if err != nil {
			log.Println(err)
			c.String(http.StatusBadRequest, "Bad request")
			return err
		}
		from = uint(from64)
	}

	param = c.QueryParam("count")
	if param != "" {
		count64, err := strconv.ParseInt(param, 10, 64)
		if err != nil {
			log.Println(err)
			c.String(http.StatusBadRequest, "Bad request")
			return err
		}
		count = uint(count64)
	}

	articles, err := a.h.GetArticles(from, count, true)
	if err != nil {
		c.String(http.StatusInternalServerError, err.Error())
		return err
	}

	c.JSON(http.StatusOK, articles)
	return nil
}

type EditArticleRequest struct {
	Title       string                  `form:"title"`
	Content     string                  `form:"content"`
	Description string                  `form:"description"`
	Publish     bool                    `form:"publish"`
	Thumbnail   *multipart.FileHeader   `form:"thumbnail"`
	Images      []*multipart.FileHeader `form:"images"`
}

func (a *ArticleAPI) PostArticle(c echo.Context) error {
	claims, err := GetJWTUserClaims(c)
	if err != nil {
		return c.String(http.StatusUnauthorized, "please login")
	}
	writerID := claims.UserID

	var request EditArticleRequest
	err = c.Bind(&request)
	if err != nil {
		return c.String(http.StatusBadRequest, "bad request")
	}

	thumbnaildb, err := a.ih.CreateImage(request.Thumbnail, util.ImageTypeThumbnail)
	if err != nil {
		return c.String(http.StatusBadRequest, err.Error())
	}
	content, err := a.ih.UploadImagesInArticle(request.Content, request.Images)
	if err != nil {
		return c.String(http.StatusBadRequest, err.Error())
	}

	err = a.h.CreateArticle(
		request.Title,
		content,
		request.Description,
		request.Publish,
		writerID,
		thumbnaildb.ID)
	if err != nil {
		return c.String(http.StatusInternalServerError, err.Error())
	}

	return c.String(http.StatusOK, "success")
}

// ?from=xxx&count=xxx
func (a *ArticleAPI) GetMyArticles(c echo.Context) error {
	claims, err := GetJWTUserClaims(c)
	if err != nil {
		return c.String(http.StatusUnauthorized, "please login")
	}
	writerID := claims.UserID

	var from, count uint = 0, 0

	param := c.QueryParam("from")
	if param != "" {
		from64, err := strconv.ParseInt(param, 10, 64)
		if err != nil {
			log.Println(err)
			c.String(http.StatusBadRequest, "Bad request")
			return err
		}
		from = uint(from64)
	}

	param = c.QueryParam("count")
	if param != "" {
		count64, err := strconv.ParseInt(param, 10, 64)
		if err != nil {
			log.Println(err)
			c.String(http.StatusBadRequest, "Bad request")
			return err
		}
		count = uint(count64)
	}

	articles, err := a.h.GetWritersArticles(writerID, false, from, count)

	if err != nil {
		return c.String(http.StatusInternalServerError, err.Error())
	}

	return c.JSON(http.StatusOK, articles)
}

func (a *ArticleAPI) PutArticle(c echo.Context) error {
	claims, err := GetJWTUserClaims(c)
	if err != nil {
		return c.String(http.StatusUnauthorized, "Please login")
	}
	writerID := claims.UserID

	id64, err := strconv.ParseUint(c.Param("id"), 10, 64)
	id := uint(id64)

	if err != nil {
		log.Println(err.Error())
		c.String(http.StatusBadRequest, "bad request")
		return err
	}

	var req EditArticleRequest

	err = c.Bind(&req)
	if err != nil {
		return c.String(http.StatusBadRequest, "bad request")
	}

	var thumbnailID uint
	if req.Thumbnail != nil {
		thumbnaildb, err := a.ih.CreateImage(req.Thumbnail, util.ImageTypeThumbnail)
		if err != nil {
			return c.String(http.StatusBadRequest, err.Error())
		}
		thumbnailID = thumbnaildb.ID
	}

	err = a.h.UpdateArticle(id, req.Title, req.Content, req.Description, req.Publish, writerID, thumbnailID)
	if err != nil {
		return c.String(http.StatusInternalServerError, err.Error())
	}

	return c.NoContent(http.StatusOK)
}

func (a *ArticleAPI) DeleteArticle(c echo.Context) error {
	claims, err := GetJWTUserClaims(c)
	if err != nil {
		return c.String(http.StatusUnauthorized, "Please login")
	}
	writerID := claims.UserID

	id64, err := strconv.ParseUint(c.Param("id"), 10, 64)
	id := uint(id64)

	if err != nil {
		log.Println(err.Error())
		c.String(http.StatusBadRequest, "Bad request")
		return err
	}

	err = a.h.DeleteArticle(id, writerID)
	if err != nil {
		return c.String(http.StatusInternalServerError, err.Error())
	}

	return c.NoContent(http.StatusOK)
}

File: frontend/src/api/article.ts
import Result from './result';
import { API, ContentTypeForm } from './api';
import { User, NewUserTemplate } from './user';
import { Image, NewImageTemplate } from './image';
import { ToFormData } from './util';

export type ArticleInfo = {
  id: number;
  created_at: string;
  updated_at: string;
  title: string;
  content: string;
  description: string;
  is_published: boolean;
  writer_id: number;
  writer: User;
  thumbnail_id: number;
  thumbnail: Image;
};

export function NewArticleInfoTemplate(): ArticleInfo {
  return {
    id: 0,
    created_at: '2006-01-02T15:04:05+09:00',
    updated_at: '2006-01-02T15:04:05+09:00',
    title: '',
    content: '',
    description: '',
    is_published: false,
    writer_id: 0,
    writer: NewUserTemplate(),
    thumbnail_id: 0,
    thumbnail: NewImageTemplate(),
  };
}

export async function GetArticle(id: string): Promise<Result<ArticleInfo>> {
  return API.GET<ArticleInfo>('/article/' + id);
}

export async function GetArticles(
  from: number,
  count: number
): Promise<Result<ArticleInfo[]>> {
  return API.GET<ArticleInfo[]>(
    '/articles?from=' + from.toString() + '&count=' + count.toString()
  );
}

export type EditArticleRequest = {
  title: string;
  content: string;
  description: string;
  publish: boolean;
  thumbnail?: File;
  images?: File[];
};

export async function PostArticle(
  req: EditArticleRequest
): Promise<Result<ArticleInfo>> {
  if (req.publish && req.thumbnail === undefined)
    return Result.InternalBadRequestError(
      'Cannot publish without any thumbnails.'
    );

  const request = ToFormData(req);
  const res = await API.POST<ArticleInfo>(
    '/auth/article',
    request,
    ContentTypeForm
  );

  return res;
}

export async function GetMyArticles(
  from: number,
  count: number
): Promise<Result<ArticleInfo[]>> {
  const res = await API.GET<ArticleInfo[]>(
    '/auth/myarticles?from=' + from.toString() + '&count=' + count.toString()
  );

  return res;
}

export async function PutArticle(
  id: number,
  req: EditArticleRequest
): Promise<Result<null>> {
  const request = ToFormData(req);

  const res = await API.PUT<null>(
    '/auth/article/' + id.toString(),
    request,
    ContentTypeForm
  );

  return res;
}

export async function DeleteArticle(id: number): Promise<Result> {
  const res = await API.DELETE('/auth/article/' + id.toString());

  return res;
}
